src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -588,47 +592,164 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
 }
 
 /*
- Return value is a "Win32 BOOL" (1 = true, 0 = false)
- */
-int32_t GlobalizationNative_StartsWith(
-                        SortHandle* pSortHandle,
-                        const UChar* lpTarget,
-                        int32_t cwTargetLength,
-                        const UChar* lpSource,
-                        int32_t cwSourceLength,
-                        int32_t options)
+ collation element is an int used for sorting. It consists of 3 components:
+    * primary - first 16 bits, representing the base letter
+    * secondary - next 8 bits, typically an accent
+    * tertiary - last 8 bits, typically the case
+
+An example (the numbers are made up to keep it simple)
+    a: 1 0 0
+    ą: 1 1 0
+    A: 1 0 1
+    Ą: 1 1 1
+
+    this method returns a mask that allows for characters comparison using specified Collator Strength
+*/
+static int32_t GetCollationElementMask(UColAttributeValue strength)
 {
-    int32_t result = FALSE;
-    UErrorCode err = U_ZERO_ERROR;
-    const UCollator* pColl = GetCollatorFromSortHandle(pSortHandle, options, &err);
+    switch (strength)
+    {
+        case UCOL_PRIMARY:
+            return UCOL_PRIMARYORDERMASK;
+        case UCOL_SECONDARY:
+            return UCOL_PRIMARYORDERMASK | UCOL_SECONDARYORDERMASK;
+        default:
+            return UCOL_PRIMARYORDERMASK | UCOL_SECONDARYORDERMASK | UCOL_TERTIARYORDERMASK;
+    }
+}
 
-    if (U_SUCCESS(err))
+static int32_t inline GetNextNonIgnorableCharacter(UCollationElements* pIterator, UErrorCode* pErrorCode)
+{
+    int32_t result = UCOL_IGNORABLE;
+
+    do
     {
-        UStringSearch* pSearch = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, &err);
-        int32_t idx = USEARCH_DONE;
+        result = ucol_next(pIterator, pErrorCode);
+    }
+    while (result == UCOL_IGNORABLE); // SimpleStartsWith_Iterators checks for error
 
-        if (U_SUCCESS(err))
+    return result;
+}
+
+static int32_t SimpleStartsWith_Iterators(UCollationElements* pPatternIterator, UCollationElements* pSourceIterator, UColAttributeValue strength)
+{
+    UErrorCode errorCode = U_ZERO_ERROR;
+
+    int32_t collationElementMask = GetCollationElementMask(strength);
+
+    while (TRUE)
+    {
+        int32_t patternItem = GetNextNonIgnorableCharacter(pPatternIterator, &errorCode);
+        if (!U_SUCCESS(errorCode))
         {
-            idx = usearch_first(pSearch, &err);
-            if (idx != USEARCH_DONE)
+            return FALSE;
+        }
+
+        int32_t sourceItem = GetNextNonIgnorableCharacter(pSourceIterator, &errorCode);
+        if (!U_SUCCESS(errorCode))
+        {
+            return FALSE;
+        }
+
+        if (patternItem == UCOL_NULLORDER)
+        {
+            if (strength >= UCOL_SECONDARY && sourceItem != UCOL_NULLORDER)
```

Comment:
do we really need the check strength >= UCOL_SECONDARY? I guess not.

Reply:
the check is outside the hot-path and I would prefer to keep it in case for some reason we use it in the future

Reply:
I am not worried about the perf,  I am worried about the future editing of the code. I'll leave it to you to decide.